app/controllers/holds_controller.rb

```diff
@@ -35,6 +37,56 @@ def new
     @bib = Bib.new(parsed_body, holdable_locations)
   end
 
+  # Handles placing holds
+  #
+  # POST /holds
+  def create
+    results = barcodes.each_with_object(success: [], error: []) do |barcode, status|
+      hold_args = { pickup_library: params['pickup_library'], pickup_by_date: params['pickup_by_date'] }
+      response = symphony_client.place_hold(patron,
+                                            current_user.session_token,
+                                            barcode,
+                                            hold_args)
+      case response.status
+      when 200
+        status[:success] << { barcode: barcode,
+                              hold_key: JSON.parse(response.body).dig('holdRecord', 'key') }
+      else
+        Rails.logger.error("Place Hold for #{barcode} by #{patron.barcode}: #{response.body}")
+        status[:error] << { barcode: barcode,
+                            error_message: JSON.parse(response.body).dig('messageList')[0].dig('message') }
+      end
+    end
+
+    session[:place_hold_results] = results
+    session[:place_hold_catkey] = params['catkey']
+
+    redirect_to result_path
+  end
+
+  # Handles place hold response
+  #
+  # GET /holds/result
+  def result
+    return redirect_to holds_path if session[:place_hold_results].blank?
+
+    @place_hold_catkey = session[:place_hold_catkey]
+    session.delete(:place_hold_catkey)
+
+    place_hold_results = session[:place_hold_results]
+    session.delete(:place_hold_results)
+
+    processed_results = place_hold_results.each do |status, results|
+      if status == 'success'
+        results.map { |result| result['placed_hold'] = hold_lookup(result['hold_key']) }
+      else
+        results.map { |result| result['failed_hold'] = item_lookup(result['barcode']) }
+      end
+    end
```

Comment:
This whole thing could be a candidate for extraction to a service object. I tried briefly, fwiw, to see if I could simplify, like 

```rb
    @place_hold_results = place_hold_results.map do |hold_result_group|
      hold_result_group.second.map {|hold_result| hold_lookup(hold_result['hold_key'])}
    end
```

but it was a dead end.

I do think you could just assign the instance variable on line 79 so it'd be 

```rb
     @place_hold_results = raw_place_hold_results.each do |status, results|
```

This would save a couple lines.

Reply:
yes, will def. refactor controller to move out more logic out.

btw committed your suggestion, good idea 👍  thanks!